source/store.ts

```diff
@@ -0,0 +1,190 @@
+// /source/store.ts
+// A `memcached` store for the `express-rate-limit` middleware.
+
+import type {
+	Store,
+	Options as RateLimitConfiguration,
+	IncrementResponse,
+} from 'express-rate-limit'
+import Memcached from 'memcached'
+import type { Options, MemcachedClient } from './types'
+
+/**
+ * The hit count and start window timestamp of a client.
+ */
+type ClientRecord = { hits: number; time: number } // [in milliseconds]
+
+/**
+ * A `Store` for the `express-rate-limit` package that stores hit counts in
+ * Memcached.
+ */
+class MemcachedStore implements Store {
+	/**
+	 * The number of milliseconds to remember that user's requests.
+	 */
+	expiration!: number
+
+	/**
+	 * The text to prepend to the key.
+	 */
+	prefix!: string
+
+	/**
+	 * The `memcached` client to use.
+	 */
+	client!: MemcachedClient
+
+	/**
+	 * @constructor for `MemcachedStore`.
+	 *
+	 * @param options {Options} - The options used to configure the store's behaviour.
+	 */
+	constructor(options?: Partial<Options>) {
+		this.prefix = options?.prefix ?? 'rl:'
+
+		if (options?.client) {
+			if (
+				typeof options.client.get === 'function' &&
+				typeof options.client.add === 'function' &&
+				typeof options.client.replace === 'function' &&
+				typeof options.client.del === 'function'
+			)
+				this.client = options.client
+			else throw new Error('An invalid memcached client was passed to store.')
+		} else this.client = new Memcached(['127.0.0.1:11211'])
+	}
+
+	/**
+	 * Method that actually initializes the store.
+	 *
+	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
+	 *
+	 * @impl
+	 */
+	init(options: RateLimitConfiguration) {
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
+	 */
+	async increment(key: string): Promise<IncrementResponse> {
+		const response = (await this.updateKey(
+			this.prefixKey(key),
+			+1,
```

Comment:
`+1` looks funny here. Would `1` also work?

```suggestion
			1,
```